### PR TITLE
Feature/uss/copy

### PR DESCRIPTION
**Pull request by KutluOzel-b**
## Proposed changes

Adds multiselect copy / paste feature to uss view 

## Release Notes

https://github.com/zowe/vscode-extension-for-zowe/issues/1549

Milestone:

Changelog:

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [ x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Updates to Documentation or Tests (if none of the other choices apply)

## Checklist

_Put an `x` in the boxes that apply. You can also fill these out after creating the PR. If you're unsure about any of them, don't hesitate to ask. We're here to help! This checklist will be used as reference for both the contributor and the reviewer_

- [ x] I have read the [CONTRIBUTOR GUIDANCE](https://github.com/zowe/vscode-extension-for-zowe/wiki/Best-Practices:-Contributor-Guidance) wiki
- [ x] PR title follows [Conventional Commits Guidelines](https://www.conventionalcommits.org/en/v1.0.0-beta.2/)
- [ x] PR Description is included
- [ ] gif or screenshot is included if visual changes are made
- [ x] `yarn workspace vscode-extension-for-zowe vscode:prepublish` has been executed
- [ ] All checks have passed (DCO, Jenkins and Code Coverage)
- [ x] I have added unit test and it is passing
- [ ] I have added integration test and it is passing
- [ ] There is coverage for the code that I have added
- [ x] I have tested it manually and there are no regressions found
- [ ] I have added necessary documentation (if appropriate)
- [ ] Any PR dependencies have been merged and published (if appropriate)

### CODE REVIEW

File: packages/zowe-explorer-api/CHANGELOG.md [hunk 1]
```diff
@@ -11,6 +11,10 @@ All notable changes to the "zowe-explorer-api" extension will be documented in t
 
 ## `2.4.0`
 
+## TBD Release
```
Comment (JillieBeanSim) [inline packages/zowe-explorer-api/CHANGELOG.md:14]: this should be above the 2.4.0 release

File: packages/zowe-explorer-api/package.json [hunk 2]
```diff
@@ -1,6 +1,6 @@
 {
   "name": "@zowe/zowe-explorer-api",
-  "version": "2.5.0-SNAPSHOT",
+  "version": "2.6.0-SNAPSHOT",
```
Comment (JillieBeanSim) [inline packages/zowe-explorer-api/package.json:3]: this needs to be reverted back to 2.5.0-SNAPSHOT

File: packages/zowe-explorer-ftp-extension/package.json [hunk 3]
```diff
@@ -47,7 +47,7 @@
     "vscode": "^1.53.2"
   },
   "dependencies": {
-    "@zowe/zowe-explorer-api": "2.5.0-SNAPSHOT",
+    "@zowe/zowe-explorer-api": "2.6.0-SNAPSHOT",
```
Comment (JillieBeanSim) [inline packages/zowe-explorer-ftp-extension/package.json:50]: same with revert to 2.5.0-SNAPSHOT

File: packages/zowe-explorer/CHANGELOG.md [hunk 4]
```diff
@@ -16,6 +16,10 @@ All notable changes to the "vscode-extension-for-zowe" extension will be documen
 
 ## `2.4.0`
 
+## TBD Release
```
Comment (JillieBeanSim) [inline packages/zowe-explorer/CHANGELOG.md:19]: move CHANGELOG above the 2.4.0 release tag

File: packages/zowe-explorer/package.json [hunk 5]
```diff
@@ -1736,7 +1776,7 @@
     "@types/promise-queue": "^2.2.0",
     "@types/selenium-webdriver": "^3.0.4",
     "@types/yargs": "^11.0.0",
-    "eslint-plugin-zowe-explorer": "2.5.0-SNAPSHOT",
+    "eslint-plugin-zowe-explorer": "2.6.0-SNAPSHOT",
```
Comment (JillieBeanSim) [inline packages/zowe-explorer/package.json:1779]: 2.5.0-SNAPSHOT

File: packages/zowe-explorer/package.json [hunk 6]
```diff
@@ -1776,7 +1816,7 @@
     "webpack-cli": "^3.3.11"
   },
   "dependencies": {
-    "@zowe/zowe-explorer-api": "2.5.0-SNAPSHOT",
+    "@zowe/zowe-explorer-api": "2.6.0-SNAPSHOT",
```
Comment (JillieBeanSim) [inline packages/zowe-explorer/package.json:1819]: 2.5.0-SNAPSHOT

**Review by JillieBeanSim — CHANGES_REQUESTED** [review]
Hey @KutluOzel-b I commented on a couple things that need to be changed

**Comment by KutluOzel-b** [comment]
@JillieBeanSim  Thanks for reviewing and feedbacks. Adressed all in : https://github.com/zowe/vscode-extension-for-zowe/pull/2004/commits/767a7038808cd1599b8f618f6e07520fd0edeb42

**Review by JillieBeanSim — CHANGES_REQUESTED** [review]
thanks for the updates @KutluOzel-b I am seeing some issues with copy/paste.  
1. copy and paste are both always in the right click menu, DS copy/paste has only copy available until something is in the clipboard then the option changes to paste.
2. uss paste fails if user only right clicks on a node then selects paste, if the node is selected then the right click menu opened and paste selected it will succeed.  In DS tree this is not the case, it will succeed if user only right clicks the node without selecting the node first.
3. user isn't presented with an option to change the file name in uss like they are with DS.  this one I am not sure if expected behavior or if we should consider allowing the rename of the copied file.

**Review by traeok — CHANGES_REQUESTED** [review]
I tried copying 3 files from my home directory to a newly-created folder: after selecting the files and clicking Copy File(s), I right-clicked on the new directory and clicked Paste. The following exception occurred after VSCode showed "Uploading copied files..." in the status bar, and it seemed to stop the rest of the logic from executing:

exception-uss-copy

I refreshed the USS panel after the exception occurred, but noticed that the directory/file structure is intact and the files were not copied into the directory.

**Comment by KutluOzel-b** [comment]
> 

@JillieBeanSim  Thanks for reviewing it. here is the solutions / discussions on the feedbacks. 
1. **copy and paste are both always in the right click menu, DS copy/paste has only copy available until something is in the clipboard then the option changes to paste.**

- I'm not sure we have control over context menu dynamically. Made some research on it and on documented features of api there isn't listed such. 

2. **uss paste fails if user only right clicks on a node then selects paste, if the node is selected then the right click menu opened and paste selected it will succeed. In DS tree this is not the case, it will succeed if user only right clicks the node without selecting the node first.**

- good catch , adressed in : c99252cabfd323e821a497e0ea33a2fb497b1a25

3. **user isn't presented with an option to change the file name in uss like they are with DS. this one I am not sure if expected behavior or if we should consider allowing the rename of the copied file.**

- since it is not like datasets ( which has to have unique identifier for all files ) I thought adding (copy) to the end of filename is better UX if there is a file with the same name under the same directory. I'm also not sure if we should provide renaming while copying. I say we should add it only for single copy and if it is asked by users.

**Comment by KutluOzel-b** [comment]
> I tried copying 3 files from my home directory to a newly-created folder: after selecting the files and clicking Copy File(s), I right-clicked on the new directory and clicked Paste. The following exception occurred after VSCode showed "Uploading copied files..." in the status bar, and it seemed to stop the rest of the logic from executing:
> 
> exception-uss-copy
> 
> I refreshed the USS panel after the exception occurred, but noticed that the directory/file structure is intact and the files were not copied into the directory.
> 
> As @JillieBeanSim suggested, I tried to paste with the directory node selected first, and also tested with just right-clicking on that node, but both methods failed to paste the files.
> 
> This _could_ be a z/OSMF-specific issue, so I will do some more testing within USS to confirm - but regardless we'll probably want to catch any exceptions that arise when trying to perform those actions.

Good catch @traeok thanks ! Added exception handler and resolved selection issue in c99252cabfd323e821a497e0ea33a2fb497b1a25

**Comment by traeok** [comment]
@KutluOzel-b I tried to copy a file and paste it into the directory that I mentioned, but received the same error - this time the exception was handled and execution stopped as expected 👍 I will do some more debugging to confirm the bug is not something on my end. I verified that I am able to copy files to that folder within a USS shell.

**Comment by KutluOzel-b** [comment]
> > 1. **copy and paste are both always in the right click menu, DS copy/paste has only copy available until something is in the clipboard then the option changes to paste.**
> > 
> > 
> > * I'm not sure we have control over context menu dynamically. Made some research on it and on documented features of api there isn't listed such.
> 
> For this I would look at how Data Sets tree is handling it, see gif attached
> 
> copy:paste [ copy:paste ](https://user-images.githubusercontent.com/49491949/206228701-189939f7-50f9-4e93-9000-d31df45abc23.gif) [ ](https://user-images.githubusercontent.com/49491949/206228701-189939f7-50f9-4e93-9000-d31df45abc23.gif)

@JillieBeanSim Thanks for the feedback. This doesn't also happens in datasets. In member view item it is always copy and in pds view Item it is always paste. Unfortunately it doesn't reacts according to clipboard contents 😞

File: packages/zowe-explorer/src/extension.ts [hunk 7]
```diff
@@ -873,7 +886,7 @@ function initSubscribers(context: vscode.ExtensionContext, theProvider: IZoweTre
     }
 }
 
-function getSelectedNodeList(node: IZoweTreeNode, nodeList: IZoweTreeNode[]): IZoweTreeNode[] {
+export function getSelectedNodeList(node: IZoweTreeNode, nodeList: IZoweTreeNode[]): IZoweTreeNode[] {
```
Comment (JillieBeanSim) [inline packages/zowe-explorer/src/extension.ts:889]: @KutluOzel-b this function seems to be a duplicate of getSelectedNodeList() in shared/utils?  We should only have one of these functions especially if they are exactly the same and I would suggest having it stay in the shared/utils file vs adding more to the extension.ts file.

**Review by JillieBeanSim — DISMISSED** [review]
while testing, if I haven't copied anything and right-click on a directory and select paste the status bar mentions uploading but nothing is actually created.  If I paste into a directory with file name already existing that I copied it automatically overwrites without notifying the user and allowing them to rename the file/folder and since the option is there even if folder is closed, this would be a good check to have.

I would still like to research if we can find a way to check the vsc clipboard and only offer paste if something is in there.

File: packages/zowe-explorer/src/uss/ZoweUSSNode.ts [hunk 8]
```diff
@@ -646,6 +646,59 @@ export class ZoweUSSNode extends ZoweTreeNode implements IZoweUSSTreeNode {
...
+            stageName: 0,
+        };
+        const options: IUploadOptions = {
+            task,
+        };
+        if (prof.profile.encoding) {
+            options.encoding = prof.profile.encoding;
+        }
+        const api = ZoweExplorerApiRegister.getUssApi(this.profile);
+        const apiResponse = await api.fileList(remotePath);
```
Comment (t1m0thyj) [inline packages/zowe-explorer/src/uss/ZoweUSSNode.ts:669]: Should we wrap this API call in a try-catch block and handle any errors?

File: packages/zowe-explorer/src/uss/ZoweUSSNode.ts [hunk 9]
```diff
@@ -646,6 +646,59 @@ export class ZoweUSSNode extends ZoweTreeNode implements IZoweUSSTreeNode {
...
+        const fileList = apiResponse.apiResponse?.items;
+
+        for (const localFile of localFileNames) {
+            try {
+                if (localFile.endsWith("/")) {
+                    const directoryPath = localFile.slice(0, -1);
+                    await api.uploadDirectory(
+                        directoryPath,
+                        remotePath + "/" + directoryPath.split("/").pop(),
```
Comment (t1m0thyj) [inline packages/zowe-explorer/src/uss/ZoweUSSNode.ts:678]: This won't work on Windows - should we use `path.basename` instead?

File: packages/zowe-explorer/src/uss/ZoweUSSNode.ts [hunk 10]
```diff
@@ -646,6 +646,59 @@ export class ZoweUSSNode extends ZoweTreeNode implements IZoweUSSTreeNode {
...
+                        options
+                    );
+                } else {
+                    let fname = localFile.split("/").pop();
```
Comment (t1m0thyj) [inline packages/zowe-explorer/src/uss/ZoweUSSNode.ts:682]: This won't work on Windows - should we use `path.basename` instead?

**Review by t1m0thyj — CHANGES_REQUESTED** [review]

File: packages/zowe-explorer/src/extension.ts (hunk 7 above)
Comment (KutluOzel-b) [reply packages/zowe-explorer/src/extension.ts:908]: @JillieBeanSim Thanks for the warning. This is caused by pulling from main branch. Keeping it in shared/utils

**Comment by traeok** [comment]
I tested these changes and I'm able to copy files now on Windows 👍 I might have found an edge case - copying an empty directory into another directory. If a folder has files inside, I can copy it into another directory, but in the case of an empty folder nothing happens. No exceptions occur, but the directory structure remains the same.

It's unlikely that a user would want to copy an empty directory into another directory when they can just right-click on the parent -> "Create Directory" - but just wanted to mention it.

**Review by traeok — APPROVED** [review]

**Review by t1m0thyj — DISMISSED** [review]
Thanks for fixing the issues with copying on Windows 🙂 

A few questions regarding the implementation:
* When you right-click on a USS directory, should the menu item say "Copy Directory" instead of "Copy Files"?
* If I have loaded directory `/a` in the USS tree that contains a subdirectory `b`, I don't see a way to create a copy of `/a/b` at the same level (`/a/b (copy)`). Should we add "Paste" to the right-click menu for the root USS node?

**Review by traeok — APPROVED** [review]

**Review by t1m0thyj — APPROVED** [review]

**Pull request merged**